xc/xenops_server_xen.ml

```diff
@@ -829,19 +829,57 @@ let _vdi_id = "vdi-id"
 
 let _dp_id = "dp-id"
 
-(* Compute the migration-safe flags *)
-let upgrade_for_migration ~xc features =
-  try
-    let msr_arch_caps = Xenctrlext.xc_get_msr_arch_caps xc in
-    let tsx_ctrl = 0x80L in
-    if Int64.(logand msr_arch_caps tsx_ctrl = tsx_ctrl) then
-      let cpuid_hle_rtm = (1 lsl 4) lor (1 lsl 11) |> Int64.of_int in
-      features.(5) <- Int64.(logor features.(5) cpuid_hle_rtm)
-    (* These still work, albeit slowly. We might prefer to hide these from
-       host's CPUID for pool leveling and guest boot purposes, but can accept it
-       during migration. *)
-  with Xenctrlext.Unix_error (Unix.ENOSYS, _) ->
-    debug "xc_get_msr_arch_caps: ENOSYS"
+(** [zeroext a b] make arrays same lengths by zeroextending on the right *)
+let zeroext a b =
```

Comment:
The bit-manipulation functions could go into a sub module `Bits`, `BitVec` or similar - you would gain something like `BitVev.diff`.

Reply:
I've created a `module BitSet` and `module Featureset`